### PR TITLE
Valhalla requires Java 24

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Valhalla requires Java 24

Added `J9_IS_CLASSFILE_OR_ROMCLASS_VALUETYPE_VERSION(classfileOrRomClass)` to determine if a class file or romclass is `valuetype` enabled;
Consolidated the checking of `VALUE_TYPES_MAJOR_VERSION` & `PREVIEW_MINOR_VERSION`;
Exclude `Access.concat(String prefix, Object value, String suffix)` before merging the latest openjdk contents.

Required by 
* https://github.com/ibmruntimes/openj9-openjdk-jdk.valuetypes/pull/13

Tested in [a personal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Pipeline_Build_Test_JDKnext_x86-64_linux_valhalla/1879/)

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: runtime/bcutil/cfreader.c [hunk 1]
```diff
@@ -3132,8 +3132,8 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 	}
 
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
-	/* Currently value type is built on JDK23, so compare with JDK23 for now. */
-	if ((flags & BCT_MajorClassFileVersionMask) < BCT_JavaMajorVersionShifted(23)) {
+	/* Currently value type is built on JDK24, so compare with JDK24 for now. */
+	if ((flags & BCT_MajorClassFileVersionMask) < BCT_JavaMajorVersionShifted(24)) {
```
Comment (theresa-m) [inline runtime/bcutil/cfreader.c:3136]: Do you mind updating and using VALUE_TYPES_MAJOR_VERSION and PREVIEW_MINOR_VERSION instead? I created these to try to minimize the number of places that need to be updated during a version change
https://github.com/eclipse-openj9/openj9/blob/master/runtime/oti/j9consts.h#L945-L946

Comment (JasonFengJ9) [reply runtime/bcutil/cfreader.c:3136]: Yup, use `J9_IS_CLASSFILE_VALUETYPE()` and also update `VALUE_TYPES_MAJOR_VERSION` to `68`.

Tried [a personal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Pipeline_Build_Test_JDKnext_x86-64_linux_valhalla/1880/)
```
10:16:00  Exception in thread "main" java.lang.ClassFormatError: JVMCFRE190 Fields of value classes must have either ACC_STATIC or ACC_FINAL flags set.; class=org/openj9/test/lworld/NullRestrictedTypeOptTests$TestStoreToNullRestrictedField, offset=785
10:16:00  	at java.base/java.lang.ClassLoader.defineClassImpl(Native Method)
10:16:00  	at java.base/java.lang.ClassLoader.defineClassInternal(ClassLoader.java:470)
10:16:00  	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:431)
10:16:00  	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
10:16:00  	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:1174)
10:16:00  	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:961)
10:16:00  	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:867)
10:16:00  	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:825)
10:16:00  	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
10:16:00  	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1099)
10:16:00  	at java.base/java.lang.Class.getDeclaredClassesImpl(Native Method)
10:16:00  	at java.base/java.lang.Class.getClasses(Class.java:605)
10:16:00  	at org.testng.internal.ClassInfoMap.registerClass(ClassInfoMap.java:43)
10:16:00  	at org.testng.internal.ClassInfoMap.<init>(ClassInfoMap.java:29)
10:16:00  	at org.testng.internal.ClassInfoMap.<init>(ClassInfoMap.java:21)
10:16:00  	at org.testng.TestRunner.initMethods(TestRunner.java:369)
10:16:00  	at org.testng.TestRunner.init(TestRunner.java:271)
10:16:00  	at org.testng.TestRunner.init(TestRunner.java:241)
10:16:00  	at org.testng.TestRunner.<init>(TestRunner.java:167)
10:16:00  	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:663)
10:16:00  	at org.testng.SuiteRunner.init(SuiteRunner.java:260)
10:16:00  	at org.testng.SuiteRunner.<init>(SuiteRunner.java:198)
10:16:00  	at org.testng.TestNG.createSuiteRunner(TestNG.java:1295)
10:16:00  	at org.testng.TestNG.createSuiteRunners(TestNG.java:1273)
10:16:00  	at org.testng.TestNG.runSuitesLocally(TestNG.java:1128)
10:16:00  	at org.testng.TestNG.runSuites(TestNG.java:1049)
10:16:00  	at org.testng.TestNG.run(TestNG.java:1017)
10:16:00  	at org.testng.TestNG.privateMain(TestNG.java:1354)
10:16:00  	at org.testng.TestNG.main(TestNG.java:1323)
10:16:00  -----------------------------------
10:16:00  NullRestrictedTypeOptTests_0_FAILED
```

Is this something related OpenJDK Valhalla update?

Comment (theresa-m) [reply runtime/bcutil/cfreader.c:3136]: I will take a look.

Comment (theresa-m) [reply runtime/bcutil/cfreader.c:3136]: The check should not be using `J9_IS_CLASSFILE_VALUETYPE` which checks some accessFlags, it should only be checking if the class version is 68 preview.

By using `J9_IS_CLASSFILE_VALUETYPE` the logic is removing the CFR_ACC_IDENTITY flag from classes it shouldn't.

Comment (JasonFengJ9) [reply runtime/bcutil/cfreader.c]: Okay, just compare with `VALUE_TYPES_MAJOR_VERSION` which matches the existing comparison with Java major version.

Comment (JasonFengJ9) [reply runtime/bcutil/cfreader.c]: Consolidated the checking of `VALUE_TYPES_MAJOR_VERSION` & `PREVIEW_MINOR_VERSION` in `J9_IS_VALUETYPE_VERSION(classfile)`.

Comment (theresa-m) [reply runtime/bcutil/cfreader.c]: Looks great, thank you!

**Review by theresa-m — APPROVED** [review]

File: runtime/oti/j9consts.h [hunk 2]
```diff
@@ -942,8 +942,9 @@ extern "C" {
 
 /* Class version minimum for value type support. */
 
-#define VALUE_TYPES_MAJOR_VERSION 67
+#define VALUE_TYPES_MAJOR_VERSION 68
 #define PREVIEW_MINOR_VERSION 65535
+#define J9_IS_VALUETYPE_VERSION(classfile) (((classfile)->majorVersion >= VALUE_TYPES_MAJOR_VERSION) && (PREVIEW_MINOR_VERSION == (classfile)->minorVersion))
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/oti/j9consts.h]: I thought it was a generic name, changed it accordingly.

File: runtime/bcutil/cfreader.c [hunk 3]
```diff
@@ -3132,24 +3132,25 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
...
-		if (J9_ARE_ANY_BITS_SET(classfile->accessFlags, CFR_ACC_IDENTITY | CFR_ACC_ENUM | CFR_ACC_MODULE)) {
-			errorCode = J9NLS_CFR_ERR_INCORRECT_FLAG_FOUND_ON_VALUE_CLASS__ID;
+				CFR_ACC_ABSTRACT | CFR_ACC_FINAL | CFR_ACC_IDENTITY | CFR_ACC_INTERFACE)
+		) {
+			errorCode = J9NLS_CFR_ERR_CLASS_MUST_HAVE_AT_LEAST_ONE_FINAL_IDENTITY_ABSTRACT_FLAG__ID;
 			offset = index - data - 2;
 			goto _errorFound;
 		}
+
+		if (J9_ARE_NO_BITS_SET((classfile)->accessFlags, CFR_ACC_IDENTITY | CFR_ACC_INTERFACE)) {
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/bcutil/cfreader.c:3145]: Added a comment.

Also exclude `Access.concat(String prefix, Object value, String suffix)` before merging the latest openjdk contents.

**Comment by hangshao0** [comment]
Jenkins test sanity,extended zlinuxval jdknext depends https://github.com/ibmruntimes/openj9-openjdk-jdk.valuetypes/pull/13

**Comment by hangshao0** [comment]
Jenkins test sanity amac jdk21

**Comment by JasonFengJ9** [comment]
Resolved the merging conflicts, keep previous PR build links:
https://openj9-jenkins.osuosl.org/job/Build_JDK21_aarch64_mac_Personal/177/
https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_sanity.functional_aarch64_mac_Personal/91/

**Pull request merged**